nixos/modules/services/networking/stubby.nix

```diff
@@ -0,0 +1,195 @@
+          # 128 is currently recommended by
+          # https://tools.ietf.org/html/draft-ietf-dprive-padding-policy-03
+        '';
+      };
+
+      idleTimeout = mkOption {
+        default = "10000";
+        type = types.str;
+        description = ''
+          EDNS0 option for keepalive idle timeout in ms as specified in
+          https://tools.ietf.org/html/rfc7828
+          This keeps idle TLS connections open to avoid the overhead of opening a new 
+          connection for every query.
+        '';
+      };
+
+      listenAddresses = mkOption {
+        default = [ "127.0.0.1" "0::1" ];
+        type = types.listOf types.str;
+        description = ''
+          Set the listen addresses for the stubby DAEMON. This specifies localhost IPv4
+          and IPv6. It will listen on port 53 by default. Use <IP_address>@<port> to 
+          specify a different port.
+          IMPORTANT NOTE: The stubby yaml parser does not recognize IPv6 addresses
+          beginning or ending with ::. Use 0:: or ::0 to avoid this.
+        '';
+      };
+
+      roundRobinUpstreams = mkOption {
+        default = "1";
+        type = types.str;
+        description = ''
+          Instructs stubby to distribute queries across all available name servers. 
+          Set to 0 to treat the upstreams below as an ordered list and use a single
+          upstream until it becomes unavailable, then use the next one.
+        '';
+      };
+
+      upstreamServers = mkOption {
+        default = "${defaultUpstream}";
+        type = types.lines;
+        description = ''
+          The list below includes all the available test servers but only has the subset
+          operated the stubby/getdns developers enabled. You can enable any of the
+          others you want to use by uncommenting the relevant section. See:
+          https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Test+Servers
+        '';
+      };
+
+      extraConfig = mkOption {
+        default = "";
+        type = types.lines;
+        description = ''
+          specify additional upstream recursive servers or additional yaml
+          formatted configuration information here.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.stubby ];
+
+     users.extraUsers.stubby = {
+       description = "stubby daemon user";
+       isSystemUser = true;
+       group = "root";
+     };
+
+    systemd.services.stubby = {
+      description = "Stubby local DNS resolver";
+      wantedBy = [ "multi-user.target" ];
+
+      preStart = ''
+        if [ ! -d ${stateDir} ]; then
+          mkdir ${stateDir}
+        fi 
+        cp ${confFile} ${stateDir}/stubby.yml
+        chown stubby:root ${stateDir}/stubby.yml
+        chmod 644 ${stateDir}/stubby.yml
+      '';
+
+      serviceConfig = {
+        ExecStart = "${pkgs.stubby}/bin/stubby -C ${stateDir}/stubby.yml";
```

Comment:
From the man page it seems that `stateDir` is not even needed actually. Just using `${pkgs.stubby}/bin/stubby -C ${confFile}` should work too, without any of the directory set up commands.